File: 32_Batch2_LegislativeAssembly/c#/Goalegislativeassembly.cs
using System;
using System.Collections.Generic;

// Abstract class providing a general template for other classes
abstract class Member
{
    protected string Name;
    protected string ContactInfo;

    protected Member(string name, string contactInfo)
    {
        Name = name;
        ContactInfo = contactInfo;
    }

    public string GetName()
    {
        return Name;
    }

    public string GetContactInfo()
    {
        return ContactInfo;
    }
}

// This class extends the Member class, inheriting its attributes and adding additional attributes like Constituency and PoliticalParty
class MLA : Member
{
    private Constituency Constituency;
    private PoliticalParty Party;

    public MLA(string name, string contactInfo, Constituency constituency, PoliticalParty party)
        : base(name, contactInfo)
    {
        Constituency = constituency;
        Party = party;
    }

    public Constituency GetConstituency()
    {
        return Constituency;
    }

    public PoliticalParty GetParty()
    {
        return Party;
    }

    public void IntroduceBill(string title, string description, Session session)
    {
        Bill bill = new Bill(title, description, this);
        Constituency.AddBill(bill);
        session.AddBill(bill);
        Console.WriteLine($"Bill introduced successfully by {Name}");
    }
}

// This class represents a Constituency, which will be used when adding and viewing MLAs and Bills
class Constituency
{
    private string Name;
    private string Area;
    private int Population;
    private List<MLA> MLAs;
    private List<Bill> Bills;

    public Constituency(string name, string area, int population)
    {
        Name = name;
        Area = area;
        Population = population;
        MLAs = new List<MLA>();
        Bills = new List<Bill>();
    }

    public void AddMLA(MLA mla)
    {
        MLAs.Add(mla);
    }

    public void AddBill(Bill bill)
    {
        Bills.Add(bill);
    }

    public string GetName()
    {
        return Name;
    }

    public List<MLA> GetMLAs()
    {
        return MLAs;
    }

    public List<Bill> GetBills()
    {
        return Bills;
    }
}

// This class represents a PoliticalParty
class PoliticalParty
{
    private string Name;
    private string Ideology;
    private string Leader;
    private List<MLA> MLAs;

    public PoliticalParty(string name, string ideology, string leader)
    {
        Name = name;
        Ideology = ideology;
        Leader = leader;
        MLAs = new List<MLA>();
    }

    public void AddMLA(MLA mla)
    {
        MLAs.Add(mla);
    }

    public string GetName()
    {
        return Name;
    }

    public List<MLA> GetMLAs()
    {
        return MLAs;
    }
}

// This class represents a Bill
class Bill
{
    private string Title;
    private string Description;
    private MLA Sponsor;
    private bool Passed;

    public Bill(string title, string description, MLA sponsor)
    {
        Title = title;
        Description = description;
        Sponsor = sponsor;
        Passed = false;
    }

    public string GetTitle()
    {
        return Title;
    }

    public string GetDescription()
    {
        return Description;
    }
}

// This class represents a Session
class Session
{
    private DateTime Date;
    private List<Bill> Bills;

    public Session(DateTime date)
    {
        Date = date;
        Bills = new List<Bill>();
    }

    public void AddBill(Bill bill)
    {
        Bills.Add(bill);
    }

    public DateTime GetDate()
    {
        return Date;
    }

    public List<Bill> GetBills()
    {
        return Bills;
    }
}

// Main driver class of the code with a menu-driven interface, uses instances of other classes
class GoaLegislativeAssembly
{
    static void Main()
    {
        Constituency constituency1 = new Constituency("Panaji", "Urban", 150000);
        Constituency constituency2 = new Constituency("Margao", "Suburban", 200000);

        PoliticalParty partyA = new PoliticalParty("Party A", "Progressive", "Leader A");
        PoliticalParty partyB = new PoliticalParty("Party B", "Conservative", "Leader B");

        MLA mla1 = new MLA("MLA 1", "1234567890", constituency1, partyA);
        MLA mla2 = new MLA("MLA 2", "9876543210", constituency2, partyB);

        constituency1.AddMLA(mla1);
        constituency2.AddMLA(mla2);

        partyA.AddMLA(mla1);
        partyB.AddMLA(mla2);

        Session session1 = new Session(DateTime.Now);
        Session session2 = new Session(DateTime.Now);

        constituency1.AddBill(new Bill("Education Reform Bill", "To improve educational standards", mla1));
        constituency2.AddBill(new Bill("Healthcare Bill", "To enhance healthcare services", mla2));

        session1.AddBill(constituency1.GetBills()[0]);
        session2.AddBill(constituency2.GetBills()[0]);

        int choice;
        do
        {
            Console.WriteLine("\nGoa Legislative Assembly Menu:");
            Console.WriteLine("1. View MLAs");
            Console.WriteLine("2. View Bills in a Session");
            Console.WriteLine("3. Introduce a Bill");
            Console.WriteLine("4. Add MLAS");
            Console.WriteLine("5. Exit");
            Console.Write("Enter your choice: ");

            choice = int.Parse(Console.ReadLine());

            switch (choice)
            {
                case 1:
                    Console.WriteLine("\nMLAs in constituency 1:");
                    foreach (MLA mla in constituency1.GetMLAs())
                    {
                        Console.WriteLine($"{mla.GetName()} - Constituency: {mla.GetConstituency().GetName()}, Party: {mla.GetParty().GetName()}");
                    }

                    Console.WriteLine("\nMLAs in constituency 2:");
                    foreach (MLA mla in constituency2.GetMLAs())
                    {
                        Console.WriteLine($"{mla.GetName()} - Constituency: {mla.GetConstituency().GetName()}, Party: {mla.GetParty().GetName()}");
                    }
                    break;

                case 2:
                    Console.WriteLine("\nBills in Session 1:");
                    foreach (Bill bill in session1.GetBills())
                    {
                        Console.WriteLine($"{bill.GetTitle()} - {bill.GetDescription()}");
                    }
                    break;

                case 3:
                    Console.Write("Enter bill title: ");
                    string title = Console.ReadLine();
                    Console.Write("Enter bill description: ");
                    string description = Console.ReadLine();
                    mla1.IntroduceBill(title, description, session1);
                    break;

                case 4:
                    Console.Write("Enter MLA name: ");
                    string mlaName = Console.ReadLine();
                    Console.Write("Enter contact info: ");
                    string contactInfo = Console.ReadLine();
                    Console.Write("Enter constituency (1 for Panaji, 2 for Margao): ");
                    int constituencyChoice = int.Parse(Console.ReadLine());
                    Console.Write("Enter party (1 for Party A, 2 for Party B): ");
                    int partyChoice = int.Parse(Console.ReadLine());

                    Constituency selectedConstituency = constituencyChoice == 1 ? constituency1 : constituency2;
                    PoliticalParty selectedParty = partyChoice == 1 ? partyA : partyB;

                    MLA newMLA = new MLA(mlaName, contactInfo, selectedConstituency, selectedParty);
                    selectedConstituency.AddMLA(newMLA);
                    selectedParty.AddMLA(newMLA);

                    Console.WriteLine("MLA added successfully.");
                    break;

                case 5:
                    Console.WriteLine("Exiting...");
                    break;

                default:
                    Console.WriteLine("Invalid choice. Please enter again.");
                    break;
            }
        } while (choice != 5);
    }
}

File: 32_Batch2_LegislativeAssembly/c++/GoaLegislativeAssembly.cpp
#include <iostream>
#include <string>
#include <vector>
#include <memory>

using namespace std;

// Forward declaration of classes
class Bill;
class Session;

// THIS IS THE ABSTRACT CLASS WHICH PROVIDES A GENERAL TEMPLATE FOR THE OTHER CLASSES
class Member {
protected:
    std::string name;
    std::string contactInfo;

public:
    Member(const std::string& name, const std::string& contactInfo)
        : name(name), contactInfo(contactInfo) {}

    virtual ~Member() = default; // Make the class polymorphic

    std::string getName() const { return name; }
    std::string getContactInfo() const { return contactInfo; }
};


// THIS CLASS REPRESENTS A BILL
class Bill {
private:
    string title;
    string description;
    shared_ptr<Member> sponsor;
    bool passed;

public:
    Bill(const string& title, const string& description, shared_ptr<Member> sponsor)
        : title(title), description(description), sponsor(sponsor), passed(false) {}

    string getTitle() const {
        return title;
    }

    string getDescription() const {
        return description;
    }
};

// THIS CLASS REPRESENTS A CONSTITUENCY
class Constituency {
private:
    string name;
    string area;
    int population;
    vector<shared_ptr<Member>> mlas;
    vector<shared_ptr<Bill>> bills;

public:
    Constituency(const string& name, const string& area, int population)
        : name(name), area(area), population(population) {}

    void addMLA(shared_ptr<Member> mla) {
        mlas.push_back(mla);
    }

    void addBill(shared_ptr<Bill> bill) {
        bills.push_back(bill);
    }

    string getName() const {
        return name;
    }

    vector<shared_ptr<Member>> getMLAs() const {
        return mlas;
    }

    vector<shared_ptr<Bill>> getBills() const {
        return bills;
    }
};

// THIS CLASS REPRESENTS A POLITICAL PARTY
class PoliticalParty {
private:
    string name;
    string ideology;
    string leader;
    vector<shared_ptr<Member>> mlas;

public:
    PoliticalParty(const string& name, const string& ideology, const string& leader)
        : name(name), ideology(ideology), leader(leader) {}

    void addMLA(shared_ptr<Member> mla) {
        mlas.push_back(mla);
    }

    string getName() const {
        return name;
    }

    vector<shared_ptr<Member>> getMLAs() const {
        return mlas;
    }
};

// THIS CLASS REPRESENTS A SESSION
class Session {
private:
    vector<shared_ptr<Bill>> bills;

public:
    Session() {}

    void addBill(shared_ptr<Bill> bill) {
        bills.push_back(bill);
    }

    vector<shared_ptr<Bill>> getBills() const {
        return bills;
    }
};

// THIS CLASS EXTENDS THE MEMBER CLASS, INHERITING ITS ATTRIBUTES AND ADDING ADDITIONAL ATTRIBUTES
class MLA : public Member, public enable_shared_from_this<MLA> {
private:
    shared_ptr<Constituency> constituency;
    shared_ptr<PoliticalParty> party;

public:
    MLA(const string& name, const string& contactInfo, shared_ptr<Constituency> constituency, shared_ptr<PoliticalParty> party)
        : Member(name, contactInfo), constituency(constituency), party(party) {}

    shared_ptr<Constituency> getConstituency() const {
        return constituency;
    }

    shared_ptr<PoliticalParty> getParty() const {
        return party;
    }

    void introduceBill(const string& title, const string& description, shared_ptr<Session> session) {
        shared_ptr<Bill> bill = make_shared<Bill>(title, description, shared_from_this());
        constituency->addBill(bill);
        session->addBill(bill);
        cout << "Bill introduced successfully by " << name << endl;
    }
};

// MAIN DRIVER CLASS
int main() {
    auto constituency1 = make_shared<Constituency>("Panaji", "Urban", 150000);
    auto constituency2 = make_shared<Constituency>("Margao", "Suburban", 200000);

    auto partyA = make_shared<PoliticalParty>("Party A", "Progressive", "Leader A");
    auto partyB = make_shared<PoliticalParty>("Party B", "Conservative", "Leader B");

    auto mla1 = make_shared<MLA>("MLA 1", "1234567890", constituency1, partyA);
    auto mla2 = make_shared<MLA>("MLA 2", "9876543210", constituency2, partyB);

    constituency1->addMLA(mla1);
    constituency2->addMLA(mla2);

    partyA->addMLA(mla1);
    partyB->addMLA(mla2);

    auto session1 = make_shared<Session>();
    auto session2 = make_shared<Session>();

    auto bill1 = make_shared<Bill>("Education Reform Bill", "To improve educational standards", mla1);
    auto bill2 = make_shared<Bill>("Healthcare Bill", "To enhance healthcare services", mla2);

    constituency1->addBill(bill1);
    constituency2->addBill(bill2);

    session1->addBill(bill1);
    session2->addBill(bill2);

    int choice;
    do {
        cout << "\nGoa Legislative Assembly Menu:" << endl;
        cout << "1. View MLAs" << endl;
        cout << "2. View Bills in a Session" << endl;
        cout << "3. Introduce a Bill" << endl;
        cout << "4. Add MLAs" << endl;
        cout << "5. Exit" << endl;
        cout << "Enter your choice: ";
        cin >> choice;

        switch (choice) {
            case 1: {
                cout << "\nMLAs in constituency 1:" << endl;
                for (const auto& member : constituency1->getMLAs()) {
                    auto mla = dynamic_pointer_cast<MLA>(member);
                    if (mla) {
                        cout << mla->getName() << " - Constituency: " << mla->getConstituency()->getName()
                             << ", Party: " << mla->getParty()->getName() << endl;
                    }
                }
                cout << "\nMLAs in constituency 2:" << endl;
                for (const auto& member : constituency2->getMLAs()) {
                    auto mla = dynamic_pointer_cast<MLA>(member);
                    if (mla) {
                        cout << mla->getName() << " - Constituency: " << mla->getConstituency()->getName()
                             << ", Party: " << mla->getParty()->getName() << endl;
                    }
                }
                break;
            }
            case 2: {
                cout << "\nBills in Session 1:" << endl;
                for (const auto& bill : session1->getBills()) {
                    cout << bill->getTitle() << " - " << bill->getDescription() << endl;
                }
                break;
            }
            case 3: {
                cin.ignore(); // Ignore the newline character left in the buffer
                string title, description;
                cout << "Enter bill title: ";
                getline(cin, title);
                cout << "Enter bill description: ";
                getline(cin, description);
                mla1->introduceBill(title, description, session1);
                break;
            }
            case 4: {
                cin.ignore(); // Ignore the newline character left in the buffer
                string mlaName, contactInfo;
                int constituencyChoice, partyChoice;

                cout << "Enter MLA name: ";
                getline(cin, mlaName);
                cout << "Enter contact info: ";
                getline(cin, contactInfo);
                cout << "Enter constituency (1 for Panaji, 2 for Margao): ";
                cin >> constituencyChoice;
                cout << "Enter party (1 for Party A, 2 for Party B): ";
                cin >> partyChoice;

                shared_ptr<Constituency> selectedConstituency = (constituencyChoice == 1) ? constituency1 : constituency2;
                shared_ptr<PoliticalParty> selectedParty = (partyChoice == 1) ? partyA : partyB;

                auto newMLA = make_shared<MLA>(mlaName, contactInfo, selectedConstituency, selectedParty);
                selectedConstituency->addMLA(newMLA);
                selectedParty->addMLA(newMLA);

                cout << "MLA added successfully." << endl;
                break;
            }
            case 5:
                cout << "Exiting..." << endl;
                break;
            default:
                cout << "Invalid choice. Please enter again." << endl;
                break;
        }
    } while (choice != 5);

    return 0;
}
